@WithoutTenant does not work on class

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Setup a MT app
2. Add `@WithoutTenant` annotation to a class

### Expected Behaviour

All methods in that class will be executed Without being Tenant aware.

### Actual Behaviour

```
Caused by: groovy.lang.MissingPropertyException: No such property: tenantId for class: mttests.WithoutController
```

### Environment Information

- **Operating System**: OSX
- **GORM Version:** 6.1.10
- **Grails Version (if using Grails):** 3.3.6
- **JDK Version:** 8u161

### Example Application

https://github.com/erichelgeson/mttests/blob/master/grails-app/controllers/mttests/WithoutController.groovy

--- a/grails-datastore-gorm-test/src/test/groovy/grails/gorm/tests/CurrentTenantTransformSpec.groovy
+++ b/grails-datastore-gorm-test/src/test/groovy/grails/gorm/tests/CurrentTenantTransformSpec.groovy
@@ -13,7 +13,6 @@ import org.grails.datastore.mapping.multitenancy.MultiTenancySettings
 import org.grails.datastore.mapping.multitenancy.exceptions.TenantNotFoundException
 import org.grails.datastore.mapping.multitenancy.resolvers.SystemPropertyTenantResolver
 import org.grails.datastore.mapping.simple.SimpleMapDatastore
-import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.transaction.TransactionStatus
 import spock.lang.AutoCleanup
 import spock.lang.Shared
@@ -36,6 +35,30 @@ class CurrentTenantTransformSpec  extends Specification {
 
     void setup() {
         System.setProperty(SystemPropertyTenantResolver.PROPERTY_NAME, "")
+    }
+
+    void "Test parsing of @WithoutTenant"() {
+        Class testServiceClass = new GroovyShell().evaluate('''
+import grails.gorm.multitenancy.WithoutTenant
+import grails.gorm.transactions.Transactional
+
+@Transactional
+@WithoutTenant
+class TestService {
+
+    Integer count() {
+        return 10;
+    }
+   
+}
+
+return TestService
+''')
+        PlayService playService = new PlayService()
+
+        expect:
+        testServiceClass.getDeclaredMethod('$mt__count')
+        playService.countPlays() == 10
     }
 
     void "Test parsing of @CurrentTenant"() {
@@ -66,7 +89,6 @@ class TestService {
 
 return TestService
 ''')
-        TeamService teamService = new TeamService()
         teamService.setTargetDatastore(Mock(Datastore))
 
         expect:
@@ -138,6 +160,14 @@ return TestService
         then:"It executed corretly without throwing an exception"
         playerCount == 0
     }
+}
+
+@WithoutTenant
+class PlayService {
+    int countPlays() {
+        return 10;
+    }
+
 }
 
 @CurrentTenant

--- a/grails-datastore-gorm-test/src/test/groovy/grails/gorm/tests/CurrentTenantTransformSpec.groovy
+++ b/grails-datastore-gorm-test/src/test/groovy/grails/gorm/tests/CurrentTenantTransformSpec.groovy
@@ -89,6 +89,7 @@ class TestService {
 
 return TestService
 ''')
+        TeamService teamService = new TeamService()
         teamService.setTargetDatastore(Mock(Datastore))
 
         expect:

--- a/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/multitenancy/transform/TenantTransform.groovy
+++ b/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/multitenancy/transform/TenantTransform.groovy
@@ -126,8 +126,8 @@ class TenantTransform extends AbstractDatastoreMethodDecoratingTransformation im
     }
 
     @Override
-    protected Parameter[] prepareNewMethodParameters(MethodNode methodNode, Map<String, ClassNode> genericsSpec) {
-        if(methodNode.getAnnotations(WITHOUT_TENANT_ANNOTATION_TYPE).isEmpty()) {
+    protected Parameter[] prepareNewMethodParameters(MethodNode methodNode, Map<String, ClassNode> genericsSpec, ClassNode classNode = null) {
+        if(methodNode.getAnnotations(WITHOUT_TENANT_ANNOTATION_TYPE).isEmpty() && (!classNode || classNode.getAnnotations(WITHOUT_TENANT_ANNOTATION_TYPE).isEmpty())) {
             final Parameter tenantIdParameter = param(make(Serializable), VAR_TENANT_ID)
             Parameter[] parameters = methodNode.getParameters()
             Parameter[] newParameters = parameters.length > 0 ? (copyParameters(((parameters as List) + [tenantIdParameter]) as Parameter[], genericsSpec)) : [tenantIdParameter] as Parameter[]
@@ -137,6 +137,7 @@ class TenantTransform extends AbstractDatastoreMethodDecoratingTransformation im
             return copyParameters(methodNode.getParameters())
         }
     }
+
     @Override
     protected boolean isValidAnnotation(AnnotationNode annotationNode, AnnotatedNode classNode) {
         ClassNode annotationClassNode = annotationNode.getClassNode()

--- a/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/transactions/transform/TransactionalTransform.groovy
+++ b/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/transactions/transform/TransactionalTransform.groovy
@@ -96,12 +96,12 @@ import static org.grails.datastore.mapping.reflect.AstUtils.*
 @CompileStatic
 @GroovyASTTransformation(phase = CompilePhase.CANONICALIZATION)
 class TransactionalTransform extends AbstractDatastoreMethodDecoratingTransformation implements Ordered {
-    private static final Set<String> ANNOTATION_NAME_EXCLUDES = new HashSet<String>([Transactional.class.getName(), "grails.transaction.Rollback", Rollback.class.getName(), NotTransactional.class.getName(), "grails.transaction.NotTransactional", "grails.gorm.transactions.ReadOnly"]);
+    private static final Set<String> ANNOTATION_NAME_EXCLUDES = new HashSet<String>([Transactional.class.getName(), "grails.transaction.Rollback", Rollback.class.getName(), NotTransactional.class.getName(), "grails.transaction.NotTransactional", "grails.gorm.transactions.ReadOnly"])
     public static final ClassNode MY_TYPE = new ClassNode(Transactional)
     public static final ClassNode READ_ONLY_TYPE = new ClassNode(ReadOnly)
     private static final String PROPERTY_TRANSACTION_MANAGER = "transactionManager"
     private static final String METHOD_EXECUTE = "execute"
-    private static final Object APPLIED_MARKER = new Object();
+    private static final Object APPLIED_MARKER = new Object()
     private static final String SET_TRANSACTION_MANAGER = "setTransactionManager"
     private static final Set<String> VALID_ANNOTATION_NAMES = Collections.unmodifiableSet(
         new HashSet<String>([Transactional.simpleName, Rollback.simpleName, ReadOnly.simpleName])
@@ -152,7 +152,7 @@ class TransactionalTransform extends AbstractDatastoreMethodDecoratingTransforma
     }
 
     @Override
-    protected Parameter[] prepareNewMethodParameters(MethodNode methodNode, Map<String, ClassNode> genericsSpec) {
+    protected Parameter[] prepareNewMethodParameters(MethodNode methodNode, Map<String, ClassNode> genericsSpec, ClassNode classNode = null) {
         final Parameter transactionStatusParameter = param(make(TransactionStatus), "transactionStatus")
         Parameter[] parameters = methodNode.getParameters()
         Parameter[] newParameters = parameters.length > 0 ? (copyParameters(((parameters as List) + [transactionStatusParameter]) as Parameter[], genericsSpec)) : [transactionStatusParameter] as Parameter[]

--- a/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/transform/AbstractMethodDecoratingTransformation.groovy
+++ b/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/transform/AbstractMethodDecoratingTransformation.groovy
@@ -48,8 +48,8 @@ import static org.grails.datastore.mapping.reflect.AstUtils.*
 @CompileStatic
 abstract class AbstractMethodDecoratingTransformation extends AbstractGormASTTransformation {
 
-    private static final Set<String> METHOD_NAME_EXCLUDES = new HashSet<String>(Arrays.asList("afterPropertiesSet", "destroy"));
-    private static final Set<String> ANNOTATION_NAME_EXCLUDES = new HashSet<String>(Arrays.asList(PostConstruct.class.getName(), PreDestroy.class.getName(), "grails.web.controllers.ControllerMethod"));
+    private static final Set<String> METHOD_NAME_EXCLUDES = new HashSet<String>(Arrays.asList("afterPropertiesSet", "destroy"))
+    private static final Set<String> ANNOTATION_NAME_EXCLUDES = new HashSet<String>(Arrays.asList(PostConstruct.class.getName(), PreDestroy.class.getName(), "grails.web.controllers.ControllerMethod"))
     /**
      * Key used to store within the original method node metadata, all previous decorated methods
      */
@@ -181,7 +181,7 @@ abstract class AbstractMethodDecoratingTransformation extends AbstractGormASTTra
             renamedMethodName = getRenamedMethodPrefix() + methodNode.getName()
         }
 
-        Parameter[] newParameters = prepareNewMethodParameters(methodNode, GenericsUtils.addMethodGenerics(methodNode, genericsSpec) )
+        Parameter[] newParameters = prepareNewMethodParameters(methodNode, GenericsUtils.addMethodGenerics(methodNode, genericsSpec), classNode)
         MethodNode renamedMethod = moveOriginalCodeToNewMethod(methodNode, renamedMethodName, newParameters, classNode, sourceUnit, genericsSpec)
         MethodCallExpression originalMethodCall = buildCallToOriginalMethod(classNode, renamedMethod)
 
@@ -226,7 +226,7 @@ abstract class AbstractMethodDecoratingTransformation extends AbstractGormASTTra
         }
     }
 
-    protected Parameter[] prepareNewMethodParameters(MethodNode methodNode, Map<String, ClassNode> genericsSpec) {
+    protected Parameter[] prepareNewMethodParameters(MethodNode methodNode, Map<String, ClassNode> genericsSpec, ClassNode classNode = null) {
         return copyParameters(methodNode.getParameters(), genericsSpec)
     }
 

